### PR TITLE
feat: ability to define a sr for each VDIs

### PR DESCRIPTION
**Pull request by badrAZ**
See vatesfr/xo-web#2070    vatesfr/xo-web#2180

### CODE REVIEW

File: src/api/vm.coffee [hunk 1]
```diff
@@ -726,13 +726,19 @@ exports.rollingDeltaBackup = rollingDeltaBackup
...
+    forEach mapVdisSrs, (srId, vdiId) =>
+      mapVdisSrsXapi[vdiId] = @getObject(srId, 'SR')._xapiId
+
+  return @importDeltaVmBackup({sr, remoteId: remote, filePath, mapVdisSrs: mapVdisSrsXapi})
 
 importDeltaBackup.params = {
   sr: { type: 'string' }
   remote: { type: 'string' }
   filePath: { type: 'string' }
+  mapVdisSr: { type: 'object', optional: true }
```
Comment (pdonias) [inline src/api/vm.coffee]: `mapVdisSrs`

File: src/xapi/index.js [hunk 2]
```diff
@@ -995,7 +996,7 @@ export default class Xapi extends XapiBase {
             [TAG_BASE_DELTA]: undefined,
             [TAG_COPY_SRC]: vdi.uuid
           },
-          sr: sr.$id
+          sr: (mapVdisSrs && mapVdisSrs[vdi.uuid]) || sr.$id
```
Comment (pdonias) [inline src/xapi/index.js]: With your implementation, won't `mapVdisSrs` always be defined?

Comment (badrAZ) [inline src/xapi/index.js]: yes, mapVdisSrs is optional. The user can choose  the default sr for all his vdis.

Comment (pdonias) [inline src/xapi/index.js]: Ok but in `vm.coffee` you create a `mapVdisSrsXapi` object even if `mapVdisSrs` is undefined. So if the user chose the default SR for all the VDIs, `mapVdisSrsXapi` will be an empty object, but not undefined.

Comment (badrAZ) [inline src/xapi/index.js]: yes but mapVdisSrsXapi[vdi.uuid] will be undefined.

File: src/xapi/index.js [hunk 3]
```diff
@@ -995,7 +996,7 @@ export default class Xapi extends XapiBase {
             [TAG_BASE_DELTA]: undefined,
             [TAG_COPY_SRC]: vdi.uuid
           },
-          sr: sr.$id
+          sr: mapVdisSrs[vdi.uuid] || sr.$id
```
Comment (julien-f) [inline src/xapi/index.js]: What if `mapVdiSrs` is `undefined`?

Comment (badrAZ) [inline src/xapi/index.js]: In API#VM mapVdisSrs receive an empty object if it is undefined.

File: src/api/vm.coffee [hunk 4]
```diff
@@ -726,13 +726,19 @@ exports.rollingDeltaBackup = rollingDeltaBackup
 
 #---------------------------------------------------------------------
 
-importDeltaBackup = ({sr, remote, filePath}) ->
-  return @importDeltaVmBackup({sr, remoteId: remote, filePath})
+importDeltaBackup = ({sr, remote, filePath, mapVdisSrs}) ->
+  mapVdisSrsXapi = {}
+
+  forEach mapVdisSrs, (srId, vdiId) =>
+    mapVdisSrsXapi[vdiId] = @getObject(srId, 'SR')._xapiId
```
Comment (julien-f) [inline src/api/vm.coffee]: I think we can skip this translation if we make it clear that `mapVdisSrs` contains only UUIDs.

@pdonias what's your opinion?

File: src/xapi/index.js [hunk 5]
```diff
@@ -924,9 +924,10 @@ export default class Xapi extends XapiBase {
   @deferrable.onFailure
   async importDeltaVm ($onFailure, delta, {
     deleteBase = false,
+    disableStartAfterImport = true,
+    mapVdisSrs = {},
     name_label = delta.vm.name_label,
-    srId = this.pool.default_SR,
-    disableStartAfterImport = true
+    srId = this.pool.default_SR
```
Comment (julien-f) [inline src/xapi/index.js]: I think this param should be optional and works even if no default SR is defined if `mapVdisSrs` is completely defined.

File: src/xo-mixins/backups.js [hunk 6]
```diff
@@ -879,7 +879,7 @@ export default class {
     }
   }
 
-  async importDeltaVmBackup ({sr, remoteId, filePath}) {
+  async importDeltaVmBackup ({sr, remoteId, filePath, mapVdisSrs}) {
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: `sr` should now be optional.

File: src/xapi/index.js [hunk 7]
```diff
@@ -947,7 +948,7 @@ export default class Xapi extends XapiBase {
       }
     }
 
-    const sr = this.getObject(srId)
+    const sr = mapVdisSrs === {} && this.getObject(srId || this.pool.default_SR)
```
Comment (julien-f) [inline src/xapi/index.js]: `mapVdisSrs === {}` cannot work, can you tell me why?

File: src/xo-mixins/backups.js [hunk 8]
```diff
@@ -879,12 +879,12 @@ export default class {
...
   }
 
-  async importDeltaVmBackup ({sr, remoteId, filePath}) {
+  async importDeltaVmBackup ({sr, remoteId, filePath, mapVdisSrs = {}}) {
     filePath = `${filePath}${DELTA_BACKUP_EXT}`
     const { datetime } = parseVmBackupPath(filePath)
 
     const handler = await this._xo.getRemoteHandler(remoteId)
-    const xapi = this._xo.getXapi(sr)
+    const xapi = this._xo.getXapi(sr || mapVdisSrs[Object.keys(mapVdisSrs)[0]])
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: This is really inefficient, here's what you could do:
```js
const getFirstKey = object => {
  for (const key in object) {
    if (Object.prototype.has.call(object, key)) {
      return key
    }
  }
}
```

File: src/xo-mixins/backups.js [hunk 9]
```diff
@@ -918,7 +918,8 @@ export default class {
 
       vm = await xapi.importDeltaVm(delta, {
         disableStartAfterImport: false,
-        srId: sr._xapiId
+        srId: sr._xapiId,
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: `sr` is no longer mandatory.

File: src/xo-mixins/backups.js [hunk 10]
```diff
@@ -129,6 +129,14 @@ async function checkFileIntegrity (handler, name) {
   //  await eventToPromise(stream, 'finish')
 }
 
+const getFirstKey = object => {
+  for (const key in object) {
+    if (Object.prototype.hasOwnProperty.call(object, key)) {
+      return key
+    }
+  }
+}
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Move this in `utils` or its own module to make it reusable.

File: src/xapi/index.js [hunk 11]
```diff
@@ -1000,7 +999,7 @@ export default class Xapi extends XapiBase {
             [TAG_BASE_DELTA]: undefined,
             [TAG_COPY_SRC]: vdi.uuid
           },
-          sr: sr.$id
+          sr: mapVdisSrs[vdi.uuid] || this.getObject(srId || this.pool.default_SR).$id
```
Comment (julien-f) [inline src/xapi/index.js]: I think it's safe to use `srId` and `this.pool.default_SR` directly.

File: src/xapi/index.js [hunk 12]
```diff
@@ -1094,6 +1093,7 @@ export default class Xapi extends XapiBase {
 
   async _migrateVmWithStorageMotion (vm, hostXapi, host, {
     migrationNetwork = find(host.$PIFs, pif => pif.management).$network, // TODO: handle not found
+    mainSr,
```
Comment (julien-f) [inline src/xapi/index.js]: `sr`.

File: src/xapi/index.js [hunk 13]
```diff
@@ -1375,6 +1378,7 @@ export default class Xapi extends XapiBase {
     const accrossPools = vm.$pool !== host.$pool
     const useStorageMotion = (
       accrossPools ||
+      mainSr ||
```
Comment (julien-f) [inline src/xapi/index.js]: All these tests should be more explicit: `!== undefined`.

File: src/xo-mixins/backups.js [hunk 14]
```diff
@@ -918,7 +919,8 @@ export default class {
 
       vm = await xapi.importDeltaVm(delta, {
         disableStartAfterImport: false,
-        srId: sr._xapiId
+        srId: sr && sr._xapiId,
```
Comment (julien-f) [inline src/xo-mixins/backups.js]: Explicit test.

File: src/xapi/index.js [hunk 15]
```diff
@@ -999,7 +999,7 @@ export default class Xapi extends XapiBase {
             [TAG_BASE_DELTA]: undefined,
             [TAG_COPY_SRC]: vdi.uuid
           },
-          sr: mapVdisSrs[vdi.uuid] || this.getObject(srId || this.pool.default_SR).$id
+          sr: mapVdisSrs[vdi.uuid] || srId || this.pool.default_SR
```
Comment (julien-f) [inline src/xapi/index.js]: Why not using `this.pool.default_SR` as a default value for `srId` in the params once and for all?

File: src/xapi/index.js [hunk 16]
```diff
@@ -1367,27 +1367,27 @@ export default class Xapi extends XapiBase {
   }
 
   async migrateVm (vmId, hostXapi, hostId, {
-    mainSr,
+    sr,
     migrationNetworkId,
-    mapVifsNetworks,
-    mapVdisSrs
+    mapVifsNetworks = {},
+    mapVdisSrs = {}
```
Comment (julien-f) [inline src/xapi/index.js]: `isEmpty` works fine with `undefined`.

File: src/xapi/index.js (hunk 15 above)
Comment (badrAZ) [inline src/xapi/index.js]: ok i will use it as a default value for srId.

**Pull request merged**